src/libcore/slice/mod.rs

```diff
@@ -5304,6 +5304,29 @@ impl<A, B> SlicePartialEq<B> for [A]
     }
 }
 
+// Use an equal-pointer optimization when types are `Eq`
+impl<A> SlicePartialEq<A> for [A]
+    where A: PartialEq<A> + Eq
+{
+    default fn equal(&self, other: &[A]) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+
+        if self.as_ptr() == other.as_ptr() {
+            return true;
+        }
+
+        for i in 0..self.len() {
```

Comment:
This could be written as:
```rust
        self.iter().zip(other.iter()).all(|(x, y)| x == y)
```

Reply:
Good point. I mindlessly copied the other impl, but as written it's not clear the bounds check for `other[i]` would always be elided. I've updated both impls to user iterators now.